### PR TITLE
Database meta data

### PR DESCRIPTION
**Pull request by fedefernandez**
This PR gives the DatabaseMetadata implementation.

Summary of the work:
- The implementation has been split in `SQLDroidDatabaseMetaData.scala` and `SQLiteDatabaseMetaDataSupport` (supports only gives boolean values indicating what's supported)
- The `PrimaryKey` logic has been moved to a `class` with an `apply` method
- `SQLDroidDatabaseMetaData` has a companion object that contains methods and fields to create SQL statements
- The cursor methods has been moved to a new object called `CursorUtils`

@juanpedromoreno please, could you review? no hurries, thanks

### CODE REVIEW

File: src/main/scala/com/fortysevendeg/android/sqlite/metadata/PrimaryKey.scala [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+  val namedPattern = """(?i).* constraint +(.*?) +primary +key *\((.*?)\).*""".r
+
+  val unnamedPattern = """(?i).* primary +key *\((.*?,+.*?)\).*""".r
+
+  def apply(connection: Connection, table: String) = {
+    val statement = connection.createStatement()
+    val pk = findInSqliteMaster(statement, table) getOrElse findInPragma(statement, table)
+    statement.close()
+    pk
```
Comment (juanpedromoreno) [inline src/main/scala/com/fortysevendeg/android/sqlite/metadata/PrimaryKey.scala]: IMHO could be handy if would have a kind of DSL for statements:

```
WithStatement {
    findInSqliteMaster(statement, table) getOrElse findInPragma(statement, table)
}
```

File: src/main/scala/com/fortysevendeg/android/sqlite/metadata/PrimaryKey.scala [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    findNamed(sqlResult) orElse findUnnamed(sqlResult)
+  }
+
+  private[this] def sqliteMaster(statement: Statement, table: String): String =
+    Option(table) match {
+      case Some(t) if !t.isEmpty =>
+        val rs = statement.executeQuery(s"SELECT sql FROM sqlite_master WHERE lower(name) = lower('${escape(t)}') AND type = 'table'")
+        val sql = if (rs.next()) Some(rs.getString(1)) else None
+        rs.close()
+        sql getOrElse (throw new SQLException(s"Table not found: '$t'"))
```
Comment (juanpedromoreno) [inline src/main/scala/com/fortysevendeg/android/sqlite/metadata/PrimaryKey.scala]: ?

```
WithRecordset(statement.executeQuery(s"SELECT sql FROM sqlite_master WHERE lower(name) = lower('${escape(t)}') AND type = 'table'")) { implicit rs =>
    if (rs.next()) Some(rs.getString(1)) else None
}
```

File: src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLDroidDatabaseMetaData.scala [hunk 3]
```diff
@@ -0,0 +1,933 @@
...
+        val columnValues = scala.Array[AnyRef](
+          javaNull, javaNull, tableName, columnName, new Integer(sqlType),
+          fieldType, javaNull, javaNull, javaNull, new Integer(10),
+          new Integer(nullable), javaNull, javaNull, javaNull, javaNull,
+          new Integer(-1), new Integer(-1), "", javaNull, javaNull,
+          javaNull, javaNull, "")
+
+        matrixCursor.addRow(columnValues)
+      }
+      Try(statement.close())
```
Comment (juanpedromoreno) [inline src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLDroidDatabaseMetaData.scala]: Does it make sense if we match this Try for logging purposes?

File: src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLDroidDatabaseMetaData.scala [hunk 4]
```diff
@@ -0,0 +1,933 @@
...
+    tableNamePattern: String
+    ): ResultSet = tablePrivilegesStatement.executeQuery()
+
+  override def getTables(
+    catalog: String,
+    schemaPattern: String,
+    tableNamePattern: String,
+    types: scala.Array[String]
+    ): ResultSet = {
+    /*
```
Comment (juanpedromoreno) [inline src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLDroidDatabaseMetaData.scala]: Do you want to make `scaladoc`? In that case you need double *.

File: src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLDroidDatabaseMetaData.scala [hunk 5]
```diff
@@ -0,0 +1,933 @@
...
+    types: scala.Array[Int]
+    ): ResultSet = udtsStatement.executeQuery()
+
+  override def getVersionColumns(
+    catalog: String,
+    schema: String,
+    table: String
+    ): ResultSet = versionColumnsStatement.executeQuery()
+
+  override def getCatalogSeparator: String = "."
```
Comment (juanpedromoreno) [inline src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLDroidDatabaseMetaData.scala]: should be this member a `val`  even being an override?

File: src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLDroidDatabaseMetaData.scala [hunk 6]
```diff
@@ -0,0 +1,933 @@
...
+
+  override def getCatalogTerm: String = "catalog"
```
Comment (juanpedromoreno) [inline src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLDroidDatabaseMetaData.scala]: Same question here and in the nexts below.

File: src/main/scala/com/fortysevendeg/android/sqlite/util/CursorUtils.scala [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    def close(resultSet: ResultSet): Unit = Try(resultSet.close())
+
+    def isClosed(resultSet: ResultSet): Boolean = resultSet.isClosed
+
+  }
+
+  private def processCursor[T, C](c: C)(process: C => T, until: Option[Int] = None)(implicit cursorProcessor: CursorProcessor[C]) = {
+    def untilTrue(size: Int) = until map (_ > size) getOrElse true
+    @tailrec
+    def loop(seq: Seq[T], c: C): Seq[T] = {
```
Comment (juanpedromoreno) [inline src/main/scala/com/fortysevendeg/android/sqlite/util/CursorUtils.scala]: Is needed `c` as a parameter in this internal function?

File: src/test/scala/com/fortysevendeg/android/sqlite/integration/SQLDroidDatabaseMetaDataSpec.scala [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+
+trait SQLDroidDatabaseMetaDataSpecification
+  extends Specification
+  with BeforeEach
+  with AfterEach
+  with Mockito {
+
+  sequential
+
+  var connection: Option[Connection] = None
```
Comment (juanpedromoreno) [inline src/test/scala/com/fortysevendeg/android/sqlite/integration/SQLDroidDatabaseMetaDataSpec.scala:20]: ;)

Comment (fedefernandez) [inline src/test/scala/com/fortysevendeg/android/sqlite/integration/SQLDroidDatabaseMetaDataSpec.scala:20]: :disappointed:

File: src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLDroidDatabaseMetaData.scala [hunk 9]
```diff
@@ -0,0 +1,933 @@
...
+import android.database.{MatrixCursor, MergeCursor}
+import com.fortysevendeg.android.sqlite.metadata.SQLDroidDatabaseMetaData._
+import com.fortysevendeg.android.sqlite.resultset.SQLDroidResultSet
+import com.fortysevendeg.android.sqlite.util.CursorUtils._
+import com.fortysevendeg.android.sqlite.{SQLDroidDriver, WrapperNotSupported, _}
+import org.sqldroid.SQLDroidConnection
+
+import scala.util.{Failure, Success, Try}
+
+class SQLDroidDatabaseMetaData(connection: Connection)
```
Comment (47degci) [inline src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLDroidDatabaseMetaData.scala:14]: Codacy Issue found: [Number of methods in class exceeds 30](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/784801318/issues/source?bid=2442814&fileBranchId=2493741#l14)

File: src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLDroidDatabaseMetaData.scala [hunk 10]
```diff
@@ -0,0 +1,933 @@
...
+    } match {
+      case Success(resultSet) if !resultSet.isClosed =>
+        val seqPos = resultSet.findColumn("SEQ")
+        val namePos = resultSet.findColumn("NAME")
+        val indexSeq: Seq[(String, Int)] = resultSet.process { rs =>
+          (rs.getString(namePos), rs.getInt(seqPos))
+        }.reverse
+
+        val indexInfoSeq = indexSeq flatMap { tuple =>
+          val (indexName, index) = tuple
```
Comment (47degci) [inline src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLDroidDatabaseMetaData.scala]: Codacy Issue found: [Field name does not match the regular expression '^[a-z][A-Za-z0-9]*$'](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/784801318/issues/source?bid=2442814&fileBranchId=2493741#l271)

File: src/test/scala/com/fortysevendeg/android/sqlite/integration/SQLDroidDatabaseMetaDataSpec.scala (hunk 8 above)
Comment (47degci) [inline src/test/scala/com/fortysevendeg/android/sqlite/integration/SQLDroidDatabaseMetaDataSpec.scala:20]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/784801332/issues/source?bid=2442814&fileBranchId=2493741#l20)

File: src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLDroidDatabaseMetaData.scala [hunk 11]
```diff
@@ -0,0 +1,933 @@
...
+  override def getCatalogs: ResultSet = catalogsStatement.executeQuery()
+
+  override def getColumnPrivileges(
+    catalog: String,
+    schema: String,
+    table: String,
+    columnNamePattern: String
+    ): ResultSet = columnPrivilegesStatement.executeQuery()
+
+  override def getColumns(
```
Comment (47degci) [inline src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLDroidDatabaseMetaData.scala:73]: Codacy Issue found: [Method is longer than 50 lines](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/784801318/issues/source?bid=2442814&fileBranchId=2493741#l71)

File: src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLiteDatabaseMetaDataSupport.scala [hunk 12]
```diff
@@ -0,0 +1,210 @@
+package com.fortysevendeg.android.sqlite.metadata
+
+import java.sql.{Connection, ResultSet, DatabaseMetaData}
+
+trait SQLiteDatabaseMetaDataSupport
```
Comment (47degci) [inline src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLiteDatabaseMetaDataSupport.scala:5]: Codacy Issue found: [Number of methods in class exceeds 30](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/784801319/issues/source?bid=2442814&fileBranchId=2493741#l5)

File: src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLiteDatabaseMetaDataSupport.scala [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+
+  override def nullsAreSortedHigh: Boolean = true
+
+  override def nullsAreSortedLow: Boolean = !nullsAreSortedHigh
+
+  override def othersDeletesAreVisible(`type`: Int): Boolean = false
+
+  override def othersInsertsAreVisible(`type`: Int): Boolean = false
+
+  override def othersUpdatesAreVisible(`type`: Int): Boolean = false
```
Comment (47degci) [inline src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLiteDatabaseMetaDataSupport.scala]: Codacy Issue found: [ type  is a reserved keyword don't use it as a name.](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/784801319/issues/source?bid=2442814&fileBranchId=2493741#l44)

File: src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLiteDatabaseMetaDataSupport.scala [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+
+  override def ownDeletesAreVisible(`type`: Int): Boolean = false
+
+  override def ownInsertsAreVisible(`type`: Int): Boolean = false
```
Comment (47degci) [inline src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLiteDatabaseMetaDataSupport.scala]: Codacy Issue found: [ type  is a reserved keyword don't use it as a name.](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/784801319/issues/source?bid=2442814&fileBranchId=2493741#l48)

File: src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLiteDatabaseMetaDataSupport.scala [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+
+  override def allTablesAreSelectable(): Boolean = true
+
+  override def autoCommitFailureClosesAllResultSets(): Boolean = false
+
+  override def dataDefinitionCausesTransactionCommit(): Boolean = false
+
+  override def dataDefinitionIgnoredInTransactions(): Boolean = false
+
+  override def deletesAreDetected(`type`: Int): Boolean = false
```
Comment (47degci) [inline src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLiteDatabaseMetaDataSupport.scala:18]: Codacy Issue found: [ type  is a reserved keyword don't use it as a name.](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/784801319/issues/source?bid=2442814&fileBranchId=2493741#l18)

File: src/main/scala/com/fortysevendeg/android/sqlite/util/CursorUtils.scala [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+      if (untilTrue(seq.size) && cursorProcessor.move(c)) loop(seq :+ process(c), c)
+      else {
+        cursorProcessor.close(c)
+        seq
+      }
+    }
+    Try(loop(Seq.empty, c)) match {
```
Comment (47degci) [inline src/main/scala/com/fortysevendeg/android/sqlite/util/CursorUtils.scala]: Codacy Issue found: [Default case not specified.](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/784801324/issues/source?bid=2442814&fileBranchId=2493741#l50)

File: src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLiteDatabaseMetaDataSupport.scala [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+
+  override def supportsOrderByUnrelated(): Boolean = false
+
+  override def supportsOuterJoins(): Boolean = true
+
+  override def supportsPositionedDelete(): Boolean = false
+
+  override def supportsPositionedUpdate(): Boolean = false
+
+  override def supportsResultSetConcurrency(`type`: Int, concurrency: Int): Boolean =
```
Comment (47degci) [inline src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLiteDatabaseMetaDataSupport.scala]: Codacy Issue found: [ type  is a reserved keyword don't use it as a name.](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/784801319/issues/source?bid=2442814&fileBranchId=2493741#l154)

File: src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLDroidDatabaseMetaData.scala (hunk 4 above)
Comment (fedefernandez) [inline src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLDroidDatabaseMetaData.scala]: It's copied from original, development notes, removing

File: src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLDroidDatabaseMetaData.scala (hunk 5 above)
Comment (fedefernandez) [inline src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLDroidDatabaseMetaData.scala]: Great!

File: src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLDroidDatabaseMetaData.scala (hunk 6 above)
Comment (fedefernandez) [inline src/main/scala/com/fortysevendeg/android/sqlite/metadata/SQLDroidDatabaseMetaData.scala]: :+1:

File: src/main/scala/com/fortysevendeg/android/sqlite/util/CursorUtils.scala (hunk 7 above)
Comment (fedefernandez) [inline src/main/scala/com/fortysevendeg/android/sqlite/util/CursorUtils.scala]: Nope, good point

**Comment by fedefernandez** [comment]
@juanpedromoreno PR ready for a second round

Thanks bro

**Comment by juanpedromoreno** [comment]
LGTM, good job!

**Pull request merged**